pkg/pool-stable-phantom/contracts/test/MockStablePhantomPool.sol

```diff
@@ -15,14 +15,46 @@
 pragma solidity ^0.7.0;
 pragma experimental ABIEncoderV2;
 
+import "@balancer-labs/v2-pool-utils/contracts/test/MockInvariantDependency.sol";
+
 import "../StablePhantomPool.sol";
 
-contract MockStablePhantomPool is StablePhantomPool {
+contract MockStablePhantomPool is StablePhantomPool, MockInvariantDependency {
     constructor(NewPoolParams memory params) StablePhantomPool(params) {
         // solhint-disable-previous-line no-empty-blocks
     }
 
     function mockCacheTokenRateIfNecessary(IERC20 token) external {
         _cacheTokenRateIfNecessary(token);
     }
+
+    function getRate() public view virtual override whenInvariantConverges returns (uint256) {
+        return super.getRate();
+    }
+
+    function _onSwapGivenIn(
+        SwapRequest memory request,
+        uint256[] memory balancesIncludingBpt,
+        uint256 indexIn,
+        uint256 indexOut
+    ) internal virtual override returns (uint256 amountOut) {
+        if (request.tokenIn != IERC20(this) && request.tokenOut != IERC20(this)) {
+            _ensureInvariantConverges();
+        }
```

Comment:
The invariant is still computed on single sided joins and exits (i.e. when token in or token out is BPT), so I'm not sure we need this `if`.

Reply:
We don't - I didn't look closely enough at the full code path (lots going on in those functions). Will just use the modifier.